RandomStateAntipattern.md

```diff
@@ -0,0 +1,383 @@
+
+```scala
+val nextLong = (l: Long) => (l * 6364136223846793005L + 1442695040888963407L, l)
+```
+
+This is an unmitigated disaster.  Let's try to add three numbers:
+
+```scala
+val sum3pof_wrong = (l: Long) => {
+  val (s, a) = nextLong(l)
+  val (ss, b) = nextLong(s)
+  val (sss, c) = nextLong(s)
+  (sss, a + b + c)
+}
+```
+
+But this is _terrible_ because we exposed ourselves to accidentally reusing seeds
+as in this buggy example (the third call should have been `nextLong(ss)`).
+
+Even worse, since we're using random numbers where the number stream and the seed
+have the same type, without extra machinery you can even get the order of arguments wrong:
+
+```scala
+val sum3pof_verywrong = (l: Long) => {
+  val (a, s) = nextLong(l)
+  val (b, ss) = nextLong(s)
+  val (c, sss) = nextLong(ss)
+  (a + b + c, sss)
+}
+```
+
+Now we've mixed up our seed and our computation, and we generated and added the
+same number three times all because the local context did not make it apparent
+which of the two results was which.  (You can avoid this by wrapping the seed
+to make it unambiguously different than the random number return type.  But
+there's no point, because the exposed states are _still_ error-prone in the
+first way; we just shouldn't do this at all.)
+
+For random numbers, where independence is (usually) everything, this is as bad
+as it gets.  Avoid this pattern like the plague!  If you must have referential
+transparency in your random numbers, use `State` or some other mechanism to
+enforce the discipline required to keep your state safely progressing!
+
+## Near-parity again: accumulation
+
+Let's suppose we want to generate more than just `Long` numbers.  Maybe we want to generate a string
+containing a random sequence of `t` and `f`.  Let's give it a go each way.
+
+First, for our mutable RNG, since we didn't bother to give `nextBool` and `nextString(length: Int)` methods, let's add them.
+
+We have a `fill` method that lets us repeat an operation a certain number of times, so we can:
+
+```scala
+implicit class MoreRandomStuff(private val r: Rng) extends AnyVal {
+  def nextBool = r.nextLong < 0
+  def nextTF = if (nextBool) 't' else 'f'
+  def nextString(length: Int) = new String(Array.fill(length)(nextTF))
+}
+```
+
+Not hard at all!
+
+Let's try with the monadic version.  We can't just use a `for` comprehension, because it doesn't support
+generating arbitrarily nested calls to `flatMap`.  Instead, we can turn to a method called `replicateA` that
+does something similar (the `A` part meaning that you're replicating just the output, not the seeds)
+
+The trick with `sequence` is that it will turn `F[G[A]]` into `G[F[A]]`, that is swap the wrapper order, assuming that
+`G` and `F` are compatible.  It turns out that `State` confuses type inference, but we can explicitly ask for what we
+want, and then it works.  Also, because we're swapping two monads, we need to build up all the operations we want inside
```

Comment:
> We are swapping two monads

Sorry for the pickiness because it's not the point of the article, but you are not swapping two monads. The fact that the outer type also has a Monad instance is not relevant to its ability to be traversed (I'm only stressing this because it's an assumption that leads to problems later).

Reply:
@SystemFw - We are swapping two monads, but that the outer type is a Monad is irrelevant to its swappability.  However, it happens to be `Traverse` _and_ Monad so the two monads get swapped.  (The two monads here are `List` and `State[Long, ?]`.)  Anyway, I'm going to cut all of this out; with `replicateA` it's unneeded.